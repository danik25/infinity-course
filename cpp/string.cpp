/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}