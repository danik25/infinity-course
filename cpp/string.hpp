/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */