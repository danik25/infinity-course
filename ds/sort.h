/*******************************************************************************
*                Data Structures - Sorting  
* file name: sort.h                                             
* Related files: sort.c sort_test.c                             
* #Version: V 1.0                                                  
* writer: Dani Benjamin                                         
* reviewer: Oded Bryt                                          
* last update: 08.05.19                                        
********************************************************************************/

#ifndef _SORT_
#define _SORT_

#include <stddef.h>  /* size_t */

/* Implementation of three basic sorting algorithm:
 * BUBBLE SORT, INSERTION SORT, SELECTION SORT.
 */  


/* The signature of the function provided by the user with the sorting rule.
 * Example for such a function decleration:
 * int IsBefore(const void *first, const void *second);
 */
typedef int (is_before_t)(const void *first, const void *second);



/* BUBBLE SORT
 *
 * Description:
 *      A simple sorting algorithm that repeatedly steps through the list,
 *      compares adjacent pairs using the IsBefore function, 
 *      and swaps them if they are in the wrong order.
 *      The pass through the list is repeated until the list is sorted.                
 *
 * Params: 
 *      array - containes the elements to be bubble-sorted.
 *      len - number of elements in the array.
 *      element_size - element size in bytes.
 *      IsBefore - the comparison function.
 *        
 * Return Value: 
 *      1 - fail
 *      0 - success
 *  
 * Time complexity:
 *      O(n^2)    
 */
int BubbleSort(void *array, size_t len, size_t element_size,
                                                          is_before_t IsBefore);


/* INSERTION SORT
 *
 * Description:
 *      A simple sorting algorithm that builds the final sorted array one item
 *      at a time. The array elements are compared with each other sequentially 
 *      using the IsBefore function, and then arranged simultaneously in some
 *      particular order.        
 *
 * Params: 
 *      array - containes the elements to be insertion-sorted.
 *      len - number of elements in the array.
 *      element_size - element size in bytes.
 *      IsBefore - the comparison function.
 *        
 * Return Value: 
 *      1 - fail
 *      0 - success
 *  
 * Time complexity:
 *      O(n^2)    
 */
int InsertionSort(void *array, size_t len, size_t element_size,
                                                          is_before_t IsBefore);


/* SELECTION SORT
 *
 * Description:
 *      A simple sorting algorithm that sorts an array by repeatedly finding the
 *      minimum element (considering the IsBefore function) from unsorted part
 *      and putting it at the beginning. This algorithm maintains two subarrays
 *      in a given array.        
 *
 * Params: 
 *      array - containes the elements to be insertion-sorted.
 *      len - number of elements in the array.
 *      element_size - element size in bytes.
 *      IsBefore - the comparison function.
 *        
 * Return Value: 
 *      1 - fail
 *      0 - success
 *  
 * Time complexity:
 *      O(n^2)    
 */
int SelectionSort(void *array, size_t len, size_t element_size, 
                                                          is_before_t IsBefore);



#endif /* _SORT_ */

