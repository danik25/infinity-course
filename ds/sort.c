/*******************************************************************************
*                Data Structures - Sorting  
* file name: sort.h                                             
* Related files: sort.c sort_test.c                             
* #Version: V 1.0                                                  
* writer: Dani Benjamin                                         
* reviewer: Oded Bryt                                          
* last update: 08.05.19                                        
********************************************************************************/
#include <stdlib.h> /* malloc */
#include <assert.h> /*assert */
#include <string.h> /* memcpy */

#include "quick_sort.h"

enum stop {GO, STOP};
enum boolean {FALSE, TRUE};
enum status {SUCCESS, FAIL};

static enum status Swap(void *ptr1, void* ptr2, size_t element_size);

int BubbleSort(void *array, size_t len, size_t element_size, 
                                                    is_before_t IsBefore)
{
    char * first = NULL;
    char * second = NULL;
    size_t len_runner = len;
    int stop_flag = GO;

    assert (array);
    assert (IsBefore);

    /* run until the end, as long that swaps are made */
    while (len > 1 && stop_flag == GO)
    {
        len_runner = len;
        first = array;
        second = first + element_size;
        stop_flag = STOP;

        while (len_runner > 1)
        {
            if (FALSE == IsBefore(first, second))
            {
                if (FAIL == Swap(first, second, element_size))
                {
                    return (1);
                }
                stop_flag = GO;
            }
            
            first += element_size;
            second += element_size;
            --len_runner;
        }

        --len;
    }

    return(0);
}

int InsertionSort(void *array, size_t len, size_t element_size, 
                                                        is_before_t IsBefore)
{
    char * runner = NULL;
    char * fixer = NULL;

    assert (array);
    assert (IsBefore);

    runner = array;

    while (len > 1)
    {
        fixer = runner + element_size;
        while (FALSE == IsBefore(fixer - element_size, fixer) && fixer != array)
        {
            /* reverse bubble */
            if (FAIL == Swap(fixer, fixer - element_size, element_size))
            {
                return (1);
            }

            fixer -= element_size;
        }

        runner += element_size;
        --len;
    }
    
    return (0);
}
  
int SelectionSort(void *array, size_t len, size_t element_size, 
                                                          is_before_t IsBefore)
{
    void * first = NULL;
    char * runner = NULL;
    size_t len_copy = len;

    assert (array);
    assert (IsBefore);

    first = array;
    runner = array;

    while (len > 1)
    {
        while (len_copy > 1)
        {
            runner += element_size;
            if(FALSE == IsBefore(first, runner))
            {
                first = runner;
            }
            --len_copy;
        }

        if (FAIL == Swap(array, first, element_size))
        {
            return (1);
        }

        array = (char*)array + element_size;
        first = runner = array;
        --len;
        len_copy = len;
    }

    return (0);
}

/*============================================================================*/

static enum status Swap(void *ptr1, void* ptr2, size_t element_size)
{
    void *tmp = (void*)malloc(element_size);

    assert (ptr1);
    assert (ptr2);

    if (NULL == tmp)
    {
        return (FAIL);
    }

    memcpy(tmp, ptr1, element_size);
    memcpy(ptr1, ptr2, element_size);
    memcpy(ptr2, tmp, element_size);

    free (tmp);
    return (SUCCESS);
}