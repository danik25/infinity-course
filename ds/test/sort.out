/home/xyz/git/dani-benjamin/ds/sort.out